### PR TITLE
Feature/bmspt 174 upgrade bcf

### PR DESCRIPTION
**Pull request by balintbendebimspot**
## Description

PR includes a tons of refactoring and restructure of codes, and a brand new feature. With the new version, the toolkit can consume a stream, and return a BCF 3.0 in memory representation regardless the specified version of the file. Technically it opens the file stream, figures out the version, builds the model accordingly, then upgrades to BCF 3.0 if needed.

* Builders are divided and refactored according to versions
* New builder functions which consume list of objects and add to existing collections
* New schema converter, which dedicated to upgrade BCF2.1 to BCF3.0

### Fixes

* [BMSPT-174](https://smino.atlassian.net/browse/BMSPT-174)
* [SMN-763](https://smino.atlassian.net/browse/SMN-763)

## Type Of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality
to not work as expected)
- [x] This change requires a documentation update
- [x] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no API changes)
- [ ] CI/Build related changes (so that Github is happy)
- [ ] Dockerfile related changes
- [ ] Documentation content changes
- [x] Tests (extended coverage or fixing existing tests)
- [ ] Updated dependencies

### API Changes

- [ ] Introducing a new method or property
- [ ] Updating the API
- [ ] Deprecating methods or properties

### Environment (Docker)

- [ ] Adding a new environment variable
- [ ] Updating the production container

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have checked my code and corrected any misspellings
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation.
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have tested my code in the local docker-compose environment
- [x] My feature branch builds on Github

### CODE REVIEW

File: src/bcf-toolkit/Builder/Bcf21/BcfBuilder.cs [hunk 1]
```diff
@@ -23,31 +20,20 @@ public class BcfBuilder : IBcfBuilder<
...
 
-  public async Task<IBcf> BuildFromStream(Stream source) {
-    _bcf.Markups = await BcfConverter.ParseMarkups<Markup, VisualizationInfo>(source);
-    _bcf.Project = await BcfConverter.ParseProject<ProjectExtension>(source);
-    return BuilderUtils.ValidateItem(_bcf);
-  }
-
-  public IBcf Build() {
+  public Bcf Build() {
     return BuilderUtils.ValidateItem(_bcf);
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf21/BcfBuilder.cs:43]: IMO it's a bit weird to have a build function which calls internally a validate function. From a reading perspective it's not clear why a validation function builds something

File: src/bcf-toolkit/Builder/Bcf21/BcfBuilderExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using BcfToolkit.Model.Bcf21;
+using BcfToolkit.Utils;
+
+namespace BcfToolkit.Builder.Bcf21;
+
+public partial class BcfBuilder {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf21/BcfBuilderExtensions.cs:9]: Maybe a more general questions is the use of so much partial classes, what's the advantage here? I found it generally harder to understand and read the classes

File: src/bcf-toolkit/Builder/Bcf21/BimSnippetBuilder.cs [hunk 3]
```diff
@@ -36,7 +38,7 @@ public class BimSnippetBuilder :
     return this;
   }
 
-  public IBimSnippet Build() {
+  public BimSnippet Build() {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf21/BimSnippetBuilder.cs:41]: Why the change to a class instead of an interface? Would be the interface not a better return type? You changed that if a few cases, why that?

File: src/bcf-toolkit/Builder/Bcf30/BcfBuilder.cs [hunk 4]
```diff
@@ -1,40 +1,46 @@
...
     var markup =
-      (Markup)BuilderUtils.BuildItem<MarkupBuilder, IMarkup>(builder);
+      BuilderUtils.BuildItem<MarkupBuilder, Markup>(builder);
     _bcf.Markups.Add(markup);
     return this;
   }
 
-  public BcfBuilder AddMarkups(List<Markup> markups) {
-    markups.ForEach(m => _bcf.Markups.Add(m));
+  public BcfBuilder SetProject(Action<ProjectBuilder> builder) {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/BcfBuilder.cs]: Would be for this not a better function name "SetProjectInfo"?

File: src/bcf-toolkit/Builder/Bcf30/BcfBuilderExtensions.cs [hunk 5]
```diff
@@ -1,23 +1,85 @@
...
+    };
+
+    // collecting commenters
+    var commenters = topic.Comments
+      .SelectMany(c => new HashSet<string> {
+        c.Author,
+        c.ModifiedAuthor
+      });
+
+    var ext = new {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/BcfBuilderExtensions.cs]: Renmae suggestion - ext -> extensions. I'm not to big of a fan from short namings

File: src/bcf-toolkit/Builder/Bcf30/ColorBuilder.cs [hunk 6]
```diff
@@ -1,10 +1,12 @@
 using System;
+using BcfToolkit.Builder.Bcf30.Interfaces;
+using BcfToolkit.Builder.Interfaces;
 using BcfToolkit.Model;
 using BcfToolkit.Model.Bcf30;
 
 namespace BcfToolkit.Builder.Bcf30;
 
-public class ColorBuilder :
+public partial class ColorBuilder :
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/ColorBuilder.cs]: Should this Builder not be called now CompoentColoringColorBuilder?

File: src/bcf-toolkit/Builder/Bcf30/ColorBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+using BcfToolkit.Model.Bcf30;
+
+namespace BcfToolkit.Builder.Bcf30;
+
+public partial class ColorBuilder {
+  public ColorBuilder AddComponents(List<Component> components) {
+    components.ForEach(_color.Components.Add);
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/ColorBuilderExtensions.cs]: Here it seems like there is something missing, like that the add function is not correclty called. But maybe I misread it

File: src/bcf-toolkit/Builder/Bcf30/CommentBuilder.cs [hunk 8]
```diff
@@ -29,7 +30,9 @@ public class CommentBuilder :
     return this;
   }
 
-  public CommentBuilder SetViewPoint(string guid) {
+  public CommentBuilder SetViewPoint(string? guid) {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/CommentBuilder.cs]: This function header also seems a bit wrong, why is it called "SetViewPoint" if you have a parameter which is called guid and internally creates a "CommentViewpoint" which sets there the GUID

File: src/bcf-toolkit/Builder/Bcf30/DocumentReferenceBuilder.cs [hunk 9]
```diff
@@ -18,13 +20,24 @@ public partial class DocumentReferenceBuilder :
     return this;
   }
 
+  public DocumentReferenceBuilder SetDocumentGuid(string? guid) {
+    if (guid != null)
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/DocumentReferenceBuilder.cs]: IMO IFs without brackets are hard to read 😅 But I guess that a code style guide thing, however we do not use that in our code.

File: src/bcf-toolkit/Builder/Bcf30/DocumentReferenceBuilder.cs [hunk 10]
```diff
@@ -18,13 +20,24 @@ public partial class DocumentReferenceBuilder :
...
+      _documentReference.DocumentGuid = guid;
+    return this;
+  }
+
+  public DocumentReferenceBuilder SetUrl(string? url) {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/DocumentReferenceBuilder.cs:28]: Not sure if this a good function design if the string is nullable than just nothing happens. Maybe it could be better to change the function to a non-nullable string and the caller needs to make sure what to do if it's null. What do you think?

File: src/bcf-toolkit/Builder/Bcf30/ExtensionsBuilderExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,40 @@
+using System.Collections.Generic;
+
+namespace BcfToolkit.Builder.Bcf30;
+
+public partial class ExtensionsBuilder {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/ExtensionsBuilderExtensions.cs:5]: Why the change to make a new file? And not put everything into one file?

File: src/bcf-toolkit/Builder/Bcf30/DocumentInfoBuilderExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,11 @@
+using System.Collections.Generic;
+using BcfToolkit.Model.Bcf30;
+
+namespace BcfToolkit.Builder.Bcf30;
+
+public partial class DocumentInfoBuilder {
+  public DocumentInfoBuilder AddDocuments(List<Document> documents) {
+    documents.ForEach(_documentInfo.Documents.Add);
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/DocumentInfoBuilderExtensions.cs:8]: This seems also not correct, is the add really working?

File: src/bcf-toolkit/Builder/Bcf30/FileBuilder.cs [hunk 13]
```diff
@@ -1,43 +1,43 @@
...
 using BcfToolkit.Model.Bcf30;
 
 namespace BcfToolkit.Builder.Bcf30;
 
-public class HeaderFileBuilder : IHeaderFileBuilder<HeaderFileBuilder> {
+public class FileBuilder : IFileBuilder<FileBuilder> {
   private readonly File _file = new();
 
-  public HeaderFileBuilder SetIfcProject(string id) {
+  public FileBuilder SetIfcProject(string id) {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/FileBuilder.cs]: rename suggestion: id => ifcProject

File: src/bcf-toolkit/Builder/Bcf30/FileBuilder.cs [hunk 14]
```diff
@@ -1,43 +1,43 @@
...
     _file.IfcProject = id;
     return this;
   }
 
-  public HeaderFileBuilder SetIfcSpatialStructureElement(string id) {
+  public FileBuilder SetIfcSpatialStructureElement(string id) {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/FileBuilder.cs]: rename suggestion: id => ifcSpatialStructureElement

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/ICommentBuilder.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+  /// <param name="comment">The comment text.</param>
+  /// <returns>Returns the builder object.</returns>
+  TBuilder SetComment(string comment);
+
+  /// <summary>
+  ///   Returns the builder object set with the `ViewPoint`.
+  /// </summary>
+  /// <param name="guid">Back reference to the viewpoint GUID.</param>
+  /// <returns>Returns the builder object.</returns>
+  TBuilder SetViewPoint(string? guid);
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/ICommentBuilder.cs]: Do we set a viewpoint here based on a guid? What if its null? IMO the function name should reflect that the view point is built via this back reference. E.g. SetViewPointFromBackReference or something like that

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IDocumentReferenceBuilder.cs [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+
+  /// <summary>
+  ///   Returns the builder object set with the `DocumentGuid`.
+  /// </summary>
+  /// <param name="guid">Guid of the referenced document.</param>
+  /// <returns>Returns the builder object.</returns>
+  TBuilder SetDocumentGuid(string guid);
+
+  /// <summary>
+  ///   Returns the builder object set with the `DocumentGuid`.
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/IDocumentReferenceBuilder.cs]: i think the documentation here is not correct

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IFileBuilder.cs [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+public interface IFileBuilder<out TBuilder> : IBuilder<File> {
+  /// <summary>
+  ///   Returns the builder object set with the `IfcProject`.
+  /// </summary>
+  /// <param name="id">
+  ///   `IfcGuid` Reference to the project to which this topic is related in
+  ///   the IFC file.
+  /// </param>
+  /// <returns>Returns the builder object.</returns>
+  TBuilder SetIfcProject(string id);
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/IFileBuilder.cs]: rename suggestion: id => ifcProject

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IFileBuilder.cs [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+
+  /// <summary>
+  ///   Returns the builder object set with the `IfcSpatialStructureElement`.
+  /// </summary>
+  /// <param name="id">
+  ///   `IfcGuid` IfcGuid Reference to the spatial structure element, e.g.
+  ///   IfcBuildingStorey, to which this topic is related.
+  /// </param>
+  /// <returns>Returns the builder object.</returns>
+  TBuilder SetIfcSpatialStructureElement(string id);
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/IFileBuilder.cs]: rename suggestion: id => ifcSpatialStructureElement

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IFileBuilder.cs [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+  ///   the IFC file.
+  /// </param>
+  /// <returns>Returns the builder object.</returns>
+  TBuilder SetIfcProject(string id);
+
+  /// <summary>
+  ///   Returns the builder object set with the `IfcSpatialStructureElement`.
+  /// </summary>
+  /// <param name="id">
+  ///   `IfcGuid` IfcGuid Reference to the spatial structure element, e.g.
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/IFileBuilder.cs]: Two times "ifcGuid" in the documentation

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IMarkupBuilder.cs [hunk 20]
```diff
@@ -0,0 +1,190 @@
...
+  /// <param name="priority">Topic priority .</param>
+  /// <returns>Returns the builder object.</returns>
+  TBuilder SetPriority(string priority);
+
+  /// <summary>
+  ///   Returns the builder object set with the `Index`.
+  /// </summary>
+  /// <param name="inx">Number to maintain the order of the topics.</param>
+  /// <returns>Returns the builder object.</returns>
+  TBuilder SetIndex(int inx);
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/IMarkupBuilder.cs]: rename suggestion: inx => index

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IMarkupBuilder.cs [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+  /// <returns>Returns the builder object.</returns>
+  TBuilder AddComment(Action<TCommentBuilder> builder);
+
+  /// <summary>
+  ///   Returns the builder object extended with `ViewPoint`.
+  /// </summary>
+  /// <param name="viewpoint">Viewpoint file name.</param>
+  /// <param name="snapshot">Snapshot file name.</param>
+  /// <param name="snapshotData">Snapshot data.</param>
+  /// <param name="index">Index.</param>
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/IMarkupBuilder.cs]: Maybe add a better documentation for index and guid if possible. I guess index is clear but the guid not so much

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IProjectBuilder.cs [hunk 22]
```diff
@@ -1,14 +1,16 @@
-using BcfToolkit.Model;
+using BcfToolkit.Builder.Interfaces;
+using BcfToolkit.Model.Bcf30;
 
-namespace BcfToolkit.Builder;
+namespace BcfToolkit.Builder.Bcf30.Interfaces;
 
-public interface IProjectBuilder<out TBuilder> : IBuilder<IProject> {
+public interface IProjectBuilder<out TBuilder> : IBuilder<ProjectInfo> {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/IProjectBuilder.cs]: Should this builder not called now IProjectInfoBuilder?

File: src/bcf-toolkit/Builder/Bcf30/MarkupBuilderExensions.cs [hunk 23]
```diff
@@ -1,8 +1,47 @@
...
+  }
+
+  public MarkupBuilder AddViewPoints(List<ViewPoint> viewpoints) {
+    viewpoints.ForEach(_markup.Topic.Viewpoints.Add);
+    return this;
+  }
+
+  public MarkupBuilder AddRelatedTopics(List<string> relatedTopicGuids) {
+    relatedTopicGuids.ForEach(id => {
+      var topic = new TopicRelatedTopicsRelatedTopic {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/MarkupBuilderExensions.cs:34]: Is this name of the class correct? TopicRelatedTopcisRelatedTopic sounds kinda wrong

File: src/bcf-toolkit/Builder/Bcf30/VisibilityBuilder.cs [hunk 24]
```diff
@@ -14,12 +15,22 @@ public partial class VisibilityBuilder :
   }
 
   public VisibilityBuilder AddException(Action<ComponentBuilder> builder) {
-    var exception =
-      (Component)BuilderUtils.BuildItem<ComponentBuilder, IComponent>(builder);
-    _visibility.Exceptions.Add(exception);
+    var component =
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/VisibilityBuilder.cs:18]: Why is this now called component instead of exception? Or is an exception a component?

File: src/bcf-toolkit/Builder/Bcf30/VisualizationInfoBuilder.cs [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+  public VisualizationInfoBuilder SetVisibility(
+    Action<VisibilityBuilder> builder) {
+    var visibility =
+      BuilderUtils
+        .BuildItem<VisibilityBuilder, ComponentVisibility>(builder);
+    _visualizationInfo.Components.Visibility = visibility;
+    return this;
+  }
+
+  public VisualizationInfoBuilder AddColoring(Action<ColorBuilder> builder) {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Builder/Bcf30/VisualizationInfoBuilder.cs]: Here the function name is maybe not totally correct. Would it not be "AddColorToColoring"? Or is that too excat in that case here? 😅 Or maybe the function should be called "AddColor"? 🤔

File: src/bcf-toolkit/Converter/Bcf21/Converter.cs [hunk 26]
```diff
@@ -6,129 +6,70 @@
...
-  /// <param name="bcf">The `IBcf` interface of the BCF.</param>
-  /// <returns></returns>
-  /// <exception cref="FileNotFoundException"></exception>
-  public async Task<Stream> BcfStream(IBcf bcf) {
+  public async Task<Stream> ToBcfStream(IBcf bcf) {
     var workingDir = Directory.GetCurrentDirectory();
     var bcfTargetPath = workingDir + "/bcf.bcfzip";
 
-    var tmpFolder = await WriteBcf(bcfTargetPath, (Bcf)bcf, false);
+    var tmpFolder = await WriteBcf((Bcf)bcf, bcfTargetPath, false);
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Converter/Bcf21/Converter.cs:73]: Will you change that code in the next PRs? That we won't have here a temp folder?

File: src/bcf-toolkit/Converter/Bcf21/SchemaConverterToBcf30.cs [hunk 27]
```diff
@@ -0,0 +1,284 @@
...
+  /// <returns>Returns the converted object.</returns>
+  public static Model.Bcf30.Bcf Convert(Model.Bcf21.Bcf from) {
+    var builder = new BcfBuilder();
+    return builder
+      .AddMarkups(from.Markups.Select(ConvertMarkup).ToList(), true)
+      .SetDocumentInfo(UpdateDocumentInfo(from.Markups
+        .SelectMany(m => m.Topic.DocumentReference)
+        .Where(r => !r.IsExternal)
+        .ToList()))
+      .SetProject(p => p.WithDefaults()) // TODO: convert project
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Converter/Bcf21/SchemaConverterToBcf30.cs]: Will you do this TODO later?

File: src/bcf-toolkit/Utils/CollectionExtensions.cs [hunk 28]
```diff
@@ -0,0 +1,10 @@
+using System.Collections.ObjectModel;
+
+namespace BcfToolkit.Utils;
+
+public static class CollectionExtensions {
+  public static void AddIfNotExists<T>(this Collection<T> collection, T item) {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Utils/CollectionExtensions.cs:6]: I saw this function on some cases where you create a new object and added some objects, where I thought if it was really necessary to use "AddIfNotExists()" or not just use "Add()". Maybe you could check the usages here

File: src/bcf-toolkit/Utils/JsonExtensions.cs [hunk 29]
```diff
@@ -11,14 +11,14 @@
...
-namespace BcfToolkit.Converter;
+namespace BcfToolkit.Utils;
 
 /// <summary>
 ///   The `JsonConverter` static class opens and parses the BCF json files
 ///   and puts their contents into the BCF models. It also writes the in
 ///   memory BCF models into json files.
 /// </summary>
-public static class JsonConverter {
+public static class JsonExtensions {
```
Comment (Mudo121Bbc) [inline src/bcf-toolkit/Utils/JsonExtensions.cs:23]: We like Regex strings as a constant, where you can give the regex also a nice name :D But also up to the used style guide. Somehow I can only make comments to diffs...

**Review by Mudo121Bbc — COMMENTED** [review]
We can also have a call about the comments if it's better :D Hope the Review helps

File: src/bcf-toolkit/Builder/Bcf21/BcfBuilder.cs (hunk 1 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf21/BcfBuilder.cs:43]: Technically, the `ValidateItem` only does the validation and returns the object. So every object is validated before it is built. I think this is the point where the user calls the `Build()`, and the object is ready to validate. Do you have an example or a recommendation which would be better from reading perspective?

File: src/bcf-toolkit/Builder/Bcf21/BcfBuilderExtensions.cs (hunk 2 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf21/BcfBuilderExtensions.cs:9]: I tried to organize the pure functions in the `BcfBuilder` in this example, but of course it is generally meant for all the builder classes as well. The pure functions mean the default methods which are set a field or add a new item to a collection for the specific model. Other extra functions are defined in the `BcfBuilderExtensions`.

File: src/bcf-toolkit/Builder/Bcf21/BimSnippetBuilder.cs (hunk 3 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf21/BimSnippetBuilder.cs:41]: Yes, I think it did not make any sense to use interface as return type, because the concrete builder must know which type is built. So for the user it is inconvenient to cast after build, plus possibility of error or wrong use.

File: src/bcf-toolkit/Builder/Bcf30/BcfBuilder.cs (hunk 4 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/BcfBuilder.cs]: Good catch, previously when there was a common interface for both versions, that made sense to implement one function, but know they are separated, i will change, thanks.

File: src/bcf-toolkit/Builder/Bcf30/BcfBuilderExtensions.cs (hunk 5 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/BcfBuilderExtensions.cs]: Make sense!

File: src/bcf-toolkit/Builder/Bcf30/BcfBuilder.cs (hunk 4 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/BcfBuilder.cs]: However, i would use the `SetProject` because in the bcf the field name is `Project`
`_bcf.Project = projectInfo;`
And we would rename the builder instead
`var projectInfo = BuilderUtils.BuildItem<ProjectInfoBuilder, ProjectInfo>(builder);`
What do you think?

File: src/bcf-toolkit/Builder/Bcf30/ColorBuilder.cs (hunk 6 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/ColorBuilder.cs]: Make sense!

File: src/bcf-toolkit/Builder/Bcf30/CommentBuilder.cs (hunk 8 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/CommentBuilder.cs]: True, i will refactor it to `SetViewPointGuid`

File: src/bcf-toolkit/Builder/Bcf30/ExtensionsBuilderExtensions.cs (hunk 11 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/ExtensionsBuilderExtensions.cs:5]: The same answer: https://github.com/bimspot/bcf-toolkit/pull/14#discussion_r1592143901

File: src/bcf-toolkit/Builder/Bcf30/DocumentInfoBuilderExtensions.cs (hunk 12 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/DocumentInfoBuilderExtensions.cs:8]: Yes this is working and I think its very cool. ForEach is a higher order function which consumes a lambda. Here in this case this is an Action<T>. Of course you could use the longer format, but we really like this functional approach.

File: src/bcf-toolkit/Builder/Bcf30/FileBuilder.cs (hunk 13 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/FileBuilder.cs]: By definition:
"`IfcGuid` Reference to the project to which this topic is related in the IFC file." :)

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/ICommentBuilder.cs (hunk 15 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/Interfaces/ICommentBuilder.cs]: This is a viewpoint guid, i renamed the function name:
`SetViewPointGuid`

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IDocumentReferenceBuilder.cs (hunk 16 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/Interfaces/IDocumentReferenceBuilder.cs]: Good catch!

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IFileBuilder.cs (hunk 19 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/Interfaces/IFileBuilder.cs]: Eagle eye! :D

File: src/bcf-toolkit/Builder/Bcf30/DocumentReferenceBuilder.cs (hunk 9 above)
Comment (DanielLepold) [reply src/bcf-toolkit/Builder/Bcf30/DocumentReferenceBuilder.cs]: it depends on what you are used to, with a 1-line condition it is more understandable for me if there is no bracket:)
this can be another solution without if:
    _documentReference.DocumentGuid = guid ?? _documentReference.DocumentGuid;

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IBcfBuilder.cs
Comment (DanielLepold) [inline src/bcf-toolkit/Builder/Bcf30/Interfaces/IBcfBuilder.cs:1]: SetExtensions method: <param name="builder">Builder of the markup.</param>  is Extensions not markup

**Review by DanielLepold — COMMENTED** [review]
Nice, awesome work, pls see my comments above.

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IBcfBuilder.cs
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/Interfaces/IBcfBuilder.cs:1]: Good catch! :D

File: src/bcf-toolkit/Utils/CollectionExtensions.cs (hunk 28 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Utils/CollectionExtensions.cs:6]: This is specifically used for creating the extensions, where all the authors, commenters, users, and topic related enums are collected. The bcf uses collection instead of HashSet, so i have to check if the values which have to be added to the collection are not duplicated.

File: src/bcf-toolkit/Converter/Bcf21/SchemaConverterToBcf30.cs (hunk 27 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Converter/Bcf21/SchemaConverterToBcf30.cs]: Yes that should be added!

File: src/bcf-toolkit/Converter/Bcf21/Converter.cs (hunk 26 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Converter/Bcf21/Converter.cs:73]: We are looking for an alternative way to not using tmp folder.

File: src/bcf-toolkit/Builder/Bcf30/VisibilityBuilder.cs (hunk 24 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/VisibilityBuilder.cs:18]: Yes, the exceptions is a collection of components
`public System.Collections.ObjectModel.Collection<Component> Exceptions`

File: src/bcf-toolkit/Builder/Bcf30/Interfaces/IProjectBuilder.cs (hunk 22 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/Interfaces/IProjectBuilder.cs]: Fixed

File: src/bcf-toolkit/Builder/Bcf30/MarkupBuilderExensions.cs (hunk 23 above)
Comment (balintbendebimspot) [reply src/bcf-toolkit/Builder/Bcf30/MarkupBuilderExensions.cs:34]: Yes because it adds a new `TopicRelatedTopcisRelatedTopic` item to the `RelatedTopics` collection.
`TopicRelatedTopcisRelatedTopic` comes from the bcf schema, but yes, its not really nice :D

File: src/bcf-toolkit/Builder/Bcf21/BcfBuilder.cs (hunk 1 above)
Comment (Mudo121Bbc) [reply src/bcf-toolkit/Builder/Bcf21/BcfBuilder.cs:43]: Okay, I'm wondering then where is the Object actually built then? Or could you just return `_bcf ` without validation and it should work too? And it is built somewhere before?

If I'm not mistaken, the current approach is just a more functional programming style, where the function returns the object. In a more OOP manner, I think this is just not necessary.

From a reading perspective, I would imagine something like this:

```
public Bcf Build() {
    BuildUtils.ValidateItem(_bcf);
    return SomeBuildFunction(_bcf);
}
```

File: src/bcf-toolkit/Builder/Bcf30/BcfBuilder.cs (hunk 4 above)
Comment (Mudo121Bbc) [reply src/bcf-toolkit/Builder/Bcf30/BcfBuilder.cs]: I think thats fine too

File: src/bcf-toolkit/Builder/Bcf30/MarkupBuilderExensions.cs (hunk 23 above)
Comment (Mudo121Bbc) [reply src/bcf-toolkit/Builder/Bcf30/MarkupBuilderExensions.cs:34]: Damn okay haha
